### PR TITLE
[wip] gossiper basic functions

### PR DESCRIPTION
**Pull request by wfystx**
- HeartbeatReceiver function is used as a server, should be run in a go routine, keep listening income heartbeat
- SendHeartBeat function is used to send heartbeat in a time interval

I'm still struggling how to build the cli to better test these functions

### CODE REVIEW

File: gossiper/gossiper.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	g.peers[peer.name] = peer
+}
+
+func (g *gossiper) RemovePeer(name string)  {
+	delete(g.peers, name)
+}
+
+func (g *gossiper) GetPeerNames() {
+	fmt.Println(g.peers)
+}
```
Comment (Bowenislandsong) [inline gossiper/gossiper.go]: Always have an extra line.

File: gossiper/gossiper.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+func (g *gossiper) AddPeer(peer gossiper)  {
+	g.peers[peer.name] = peer
+}
+
+func (g *gossiper) RemovePeer(name string)  {
+	delete(g.peers, name)
+}
+
+func (g *gossiper) GetPeerNames() {
```
Comment (Bowenislandsong) [inline gossiper/gossiper.go]:
```suggestion
func (g *gossiper) PrintPeerNames() {
```

File: gossiper/gossiper.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package gossiper
+
+import "fmt"
+
+type gossiper struct {
+	name	string
+	ip		string
+	peers	map[string]gossiper
```
Comment (Bowenislandsong) [inline gossiper/gossiper.go]: Might want a flat map? IMO, we do not need the entire network information. Wdyt?

File: gossiper/gossiper_test.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package gossiper
+
+import "testing"
+
+func TestAddRemovePeer(t *testing.T)  {
+	g := new(gossiper)
+	g.name = "test"
+	g.AddPeer(gossiper{"peer", "localhost", make(map[string]gossiper)})
+	if _, ok := g.peers["peer"]; !ok {
+		t.Errorf("Add Failed")
```
Comment (Bowenislandsong) [inline gossiper/gossiper_test.go]: Might wanna look at testing package -> https://godoc.org/github.com/stretchr/testify/assert and https://godoc.org/github.com/stretchr/testify/require

**Review by Bowenislandsong — CHANGES_REQUESTED** [review]
Good initial work. A better way maybe to draw out what you want to implement as a plan. Else this could go for a bit long.

**Comment by Bowenislandsong** [comment]
Your title [wip] is enough to say it is work in progress. Your actual commit message should include the motivation of the work, what the implementation does, and why it would fit the profile.

File: gossiper/gossiper.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package gossiper
```
Comment (Bowenislandsong) [inline gossiper/gossiper.go]: pkg/gossiper/peers.go

File: gossiper/gossiper.go (hunk 3 above)
Comment (wfystx) [reply gossiper/gossiper.go]: Yeah we don't need the entire network information, I was thinking each node uses this map to keep its peers

File: gossiper/gossiper_test.go (hunk 4 above)
Comment (wfystx) [reply gossiper/gossiper_test.go]: will do

File: go.mod [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module MyGossip
+
+go 1.14
+
+replace github.com/Never-M/MyGossip => /Users/wangfuyao/go/src/github.com/Never-M/MyGossip
```
Comment (Bowenislandsong) [inline go.mod]: Why is this necessary?

File: go.mod [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+require (
+	github.com/Never-M/MyGossip v0.0.0-00010101000000-000000000000 // indirect
```
Comment (Bowenislandsong) [inline go.mod]: Also don’t think this is necessary

File: pkg/gossiper/peers.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (g *gossiper) SendHeartBeat() {
+	ctx := context.WithValue(context.Background(), "heartbeat", &heartBeat{g.name, time.Now()})
+	ctx = context.WithValue(ctx, "key", "value")
+	req, err := http.NewRequestWithContext(ctx, "GET", "http://localhost:8085/heartbeat", nil)
+	if err != nil {
+		log.Println(err)
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: Use log.info() from logrus https://github.com/sirupsen/logrus#example

File: go.mod (hunk 7 above)
Comment (Bowenislandsong) [reply go.mod]: Could just let it generate itself with go mod init

File: pkg/gossiper/peers.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	peers	map[string]gossiper
+	client  *http.Client
+}
+
+type heartBeat struct {
+	gossiperName	string
+	currentTime		time.Time
+}
+
+func NewGossiper(self string, address string) *gossiper {
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: (self, address string)

Also could just name self `name`

File: pkg/gossiper/peers.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+	g := &gossiper{
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: Return &gossiper{...

File: pkg/gossiper/peers.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		name: self,
+		ip:address,
+		peers:make(map[string]gossiper),
+		client:&http.Client{},
+	}
+	return g;
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: Don’t need `;`

File: pkg/gossiper/peers.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+func (g *gossiper) AddPeer(peer gossiper)  {
+	g.peers[peer.name] = peer
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: Might wanna check for dup

File: pkg/gossiper/peers.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+		log.Println(err)
+	}
+}
+
+func (g *gossiper) HeartBeatReceiver() {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/heartbeat", g.HeartBeatHandler)
+	contextedMux := g.HeartBeatContext(mux)
+	log.Println("Start server on port :8085")
+	http.ListenAndServe(":8085", contextedMux)
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers.go]: Should make this at least a variable if not a global var

File: pkg/gossiper/peers_test.go [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+func TestAddRemovePeer(t *testing.T)  {
+	g := new(gossiper)
+	g.name = "test"
+	g.AddPeer(gossiper{"peer", "localhost", make(map[string]gossiper)})
+	if _, ok := g.peers["peer"]; !ok {
+		t.Errorf("Add Failed")
+	}
+	g.RemovePeer("peer")
+	if _, ok := g.peers["peer"]; ok {
+		t.Errorf("Remove Failed")
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers_test.go]: Try assert package from testify and it would be like Assert.False(t,ok)

File: pkg/gossiper/peers_test.go [hunk 15]
```diff
@@ -0,0 +1,16 @@
+package gossiper
+
+import "testing"
+
+func TestAddRemovePeer(t *testing.T)  {
+	g := new(gossiper)
```
Comment (Bowenislandsong) [inline pkg/gossiper/peers_test.go]: Shouldn’t this be NewGossiper(“test”,...)?

**Review by Bowenislandsong — CHANGES_REQUESTED** [review]
Some initial comments. This is kinda hard to do on an iPad with the github app. I do not recommend using the github app, not yet. It can’t code or review, I don’t know what it’s good for. 🤣🤣🤣

File: pkg/gossiper/peers.go (hunk 8 above)
Comment (wfystx) [reply pkg/gossiper/peers.go]: Yes that's what I thought too. @Never-M can you add logrus to our project?

File: go.mod (hunk 6 above)
Comment (wfystx) [reply go.mod]: If I don't do this, it would cause a package not found error. I haven't found a solution for it

File: pkg/gossiper/db.go [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+	db   *leveldb.DB
+}
+
+type pair struct {
+	key string
+	val string
+}
+
+//how to use return &mydb{} method？
+func newdb(path string) (int, *mydb) {
```
Comment (wfystx) [inline pkg/gossiper/db.go:19]: Capital

File: pkg/gossiper/db.go [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+	newdb.path = path
+	var err error
+	newdb.db, err = leveldb.OpenFile("path/to/db", nil)
+	if err != nil {
+		return DB_CREATE_ERROR, nil
+	}
+	return SUCCEED, newdb
+}
+
+func (d *mydb) put(key, val string) int {
```
Comment (wfystx) [inline pkg/gossiper/db.go:30]: Capital
